File: scripts/migration/process-images.py
"""
Process Product Images: Remove Background & Add Shadow

This script:
1. Fetches all product images from Strapi
2. Removes backgrounds using rembg (AI-powered)
3. Adds a soft drop shadow
4. Re-uploads processed images to Strapi

Usage:
    python scripts/migration/process-images.py

Required environment variables:
    STRAPI_URL - Strapi base URL (default: http://localhost:1337)
    STRAPI_ADMIN_TOKEN - Strapi API token with write access
"""

import os
import sys
import time
import requests
from io import BytesIO
from pathlib import Path

# Load .env.local
env_path = Path(__file__).parent.parent.parent / 'apps' / 'web' / '.env.local'
if env_path.exists():
    with open(env_path, 'r') as f:
        for line in f:
            line = line.strip()
            if '=' in line and not line.startswith('#'):
                key, value = line.split('=', 1)
                os.environ[key.strip()] = value.strip()

STRAPI_URL = os.environ.get('STRAPI_URL', 'http://localhost:1337')
STRAPI_TOKEN = os.environ.get('STRAPI_ADMIN_TOKEN')

# Lazy imports for heavy libraries
rembg = None
Image = None
ImageFilter = None
ImageDraw = None

def init_libraries():
    """Initialize heavy libraries on first use."""
    global rembg, Image, ImageFilter, ImageDraw
    if rembg is None:
        print("Loading AI model (first run may take a minute)...")
        from rembg import remove
        from PIL import Image as PILImage, ImageFilter as PILFilter, ImageDraw as PILDraw
        rembg = remove
        Image = PILImage
        ImageFilter = PILFilter
        ImageDraw = PILDraw


def add_shadow(img, offset=(8, 8), shadow_color=(0, 0, 0, 80), blur_radius=15, padding=30):
    """Add a soft drop shadow to a transparent PNG image."""
    # Ensure RGBA
    if img.mode != 'RGBA':
        img = img.convert('RGBA')

    # Create new canvas with padding for shadow
    new_width = img.width + padding * 2 + abs(offset[0])
    new_height = img.height + padding * 2 + abs(offset[1])

    # Create shadow layer
    shadow = Image.new('RGBA', (new_width, new_height), (255, 255, 255, 0))

    # Get alpha channel from original image for shadow shape
    alpha = img.split()[3]
    shadow_shape = Image.new('RGBA', img.size, shadow_color)
    shadow_shape.putalpha(alpha)

    # Paste shadow with offset
    shadow_x = padding + offset[0]
    shadow_y = padding + offset[1]
    shadow.paste(shadow_shape, (shadow_x, shadow_y))

    # Blur the shadow
    for _ in range(3):  # Multiple passes for smoother blur
        shadow = shadow.filter(ImageFilter.GaussianBlur(blur_radius // 3))

    # Create final image with white background
    final = Image.new('RGBA', (new_width, new_height), (255, 255, 255, 255))

    # Composite shadow onto white background
    final = Image.alpha_composite(final, shadow)

    # Paste original image on top
    img_x = padding
    img_y = padding
    final.paste(img, (img_x, img_y), img)

    # Convert to RGB for JPEG output (optional, keeps as PNG for transparency)
    return final


def get_all_strapi_files():
    """Fetch all image files from Strapi."""
    print("Fetching files from Strapi...")

    response = requests.get(
        f"{STRAPI_URL}/api/upload/files",
        headers={"Authorization": f"Bearer {STRAPI_TOKEN}"}
    )

    if not response.ok:
        raise Exception(f"Failed to fetch files: {response.status_code}")

    all_files = response.json()

    # Filter to only original images (exclude thumbnails)
    images = [
        f for f in all_files
        if f.get('mime', '').startswith('image/')
        and not f.get('hash', '').startswith(('thumbnail_', 'small_', 'medium_', 'large_', 'xsmall_'))
    ]

    print(f"  Found {len(images)} original images")
    return images


def get_products_with_images():
    """Get all products that have images."""
    print("Fetching products with images...")
    products = []
    page = 1

    while True:
        response = requests.get(
            f"{STRAPI_URL}/api/products?populate=images&pagination[page]={page}&pagination[pageSize]=100",
            headers={"Authorization": f"Bearer {STRAPI_TOKEN}"}
        )

        if not response.ok:
            raise Exception(f"Failed to fetch products: {response.status_code}")

        data = response.json()
        if not data.get('data'):
            break

        for product in data['data']:
            if product.get('images') and len(product['images']) > 0:
                products.append({
                    'id': product['id'],
                    'documentId': product['documentId'],
                    'name': product['name'],
                    'image': product['images'][0]
                })

        if len(data['data']) < 100:
            break
        page += 1

    print(f"  Found {len(products)} products with images")
    return products


def download_image(url):
    """Download image from URL."""
    if url.startswith('/'):
        url = f"{STRAPI_URL}{url}"

    response = requests.get(url, timeout=30)
    if not response.ok:
        return None

    return BytesIO(response.content)


def process_image(image_bytes):
    """Remove background and add shadow."""
    init_libraries()

    # Open image
    img = Image.open(image_bytes)

    # Remove background
    output = rembg(img)

    # Add shadow
    result = add_shadow(output)

    return result


def upload_image_to_strapi(img, filename):
    """Upload processed image to Strapi."""
    # Save to bytes
    buffer = BytesIO()

    # Save as PNG to preserve transparency, or JPEG for smaller size
    if filename.lower().endswith('.png'):
        img.save(buffer, format='PNG', optimize=True)
    else:
        # Convert to RGB for JPEG
        if img.mode == 'RGBA':
            rgb_img = Image.new('RGB', img.size, (255, 255, 255))
            rgb_img.paste(img, mask=img.split()[3])
            img = rgb_img
        img.save(buffer, format='JPEG', quality=90, optimize=True)

    buffer.seek(0)

    # Upload to Strapi
    files = {'files': (filename, buffer, 'image/png' if filename.endswith('.png') else 'image/jpeg')}

    response = requests.post(
        f"{STRAPI_URL}/api/upload",
        headers={"Authorization": f"Bearer {STRAPI_TOKEN}"},
        files=files
    )

    if not response.ok:
        return None

    result = response.json()
    if result and len(result) > 0:
        return result[0]['id']
    return None


def update_product_image(document_id, file_id):
    """Update product to use new image."""
    response = requests.put(
        f"{STRAPI_URL}/api/products/{document_id}",
        headers={
            "Authorization": f"Bearer {STRAPI_TOKEN}",
            "Content-Type": "application/json"
        },
        json={"data": {"images": [file_id]}}
    )
    return response.ok


def delete_old_file(file_id):
    """Delete old file from Strapi."""
    try:
        requests.delete(
            f"{STRAPI_URL}/api/upload/files/{file_id}",
            headers={"Authorization": f"Bearer {STRAPI_TOKEN}"}
        )
    except:
        pass


def main():
    print("=" * 50)
    print("  Image Background Removal & Shadow")
    print("=" * 50)
    print()

    if not STRAPI_TOKEN:
        print("ERROR: STRAPI_ADMIN_TOKEN is required")
        sys.exit(1)

    # Get products with images
    products = get_products_with_images()

    if not products:
        print("No products with images found!")
        return

    print(f"\nProcessing {len(products)} product images...\n")

    processed = 0
    failed = 0

    for i, product in enumerate(products):
        try:
            image = product['image']
            image_url = image.get('url', '')
            image_name = image.get('name', 'image.png')
            old_file_id = image.get('id')

            # Download
            image_bytes = download_image(image_url)
            if not image_bytes:
                print(f"  [SKIP] {product['name']}: download failed")
                failed += 1
                continue

            # Process (remove bg + add shadow)
            result_img = process_image(image_bytes)

            # Generate new filename
            base_name = Path(image_name).stem
            new_filename = f"{base_name}_processed.png"

            # Upload new image
            new_file_id = upload_image_to_strapi(result_img, new_filename)
            if not new_file_id:
                print(f"  [FAIL] {product['name']}: upload failed")
                failed += 1
                continue

            # Update product
            success = update_product_image(product['documentId'], new_file_id)
            if not success:
                print(f"  [FAIL] {product['name']}: link failed")
                failed += 1
                continue

            # Delete old file
            if old_file_id:
                delete_old_file(old_file_id)

            processed += 1
            print(f"  [{processed}/{len(products)}] {product['name']}")

            # Progress update
            if processed % 10 == 0:
                print(f"\n  Progress: {processed}/{len(products)} ({failed} failed)\n")

            # Small delay to avoid overloading
            time.sleep(0.1)

        except Exception as e:
            print(f"  [ERROR] {product['name']}: {str(e)}")
            failed += 1

    print()
    print("=" * 50)
    print(f"  Processed: {processed}")
    print(f"  Failed: {failed}")
    print("=" * 50)


if __name__ == "__main__":
    main()

File: scripts/migration/process-images.ts
/**
 * Process Product Images: Remove Background & Add Shadow
 *
 * This script:
 * 1. Fetches all product images from Strapi
 * 2. Removes backgrounds using @imgly/background-removal-node
 * 3. Adds a soft drop shadow using Sharp
 * 4. Re-uploads processed images to Strapi
 *
 * Usage:
 *   npx tsx scripts/migration/process-images.ts
 */

import fs from 'fs';
import path from 'path';
import sharp from 'sharp';

// Load .env.local
const envPath = path.join(__dirname, '../../apps/web/.env.local');
if (fs.existsSync(envPath)) {
  const envContent = fs.readFileSync(envPath, 'utf-8');
  for (const line of envContent.split('\n')) {
    const [key, ...valueParts] = line.split('=');
    if (key && valueParts.length > 0) {
      process.env[key.trim()] = valueParts.join('=').trim();
    }
  }
}

const STRAPI_URL = process.env.STRAPI_URL || 'http://localhost:1337';
const STRAPI_TOKEN = process.env.STRAPI_ADMIN_TOKEN;

// Will be initialized lazily
let removeBackground: (input: Blob) => Promise<Blob>;

async function initBackgroundRemoval() {
  console.log('Loading AI model (first run downloads ~180MB)...');
  const { removeBackground: rb } = await import('@imgly/background-removal-node');
  removeBackground = rb;
  console.log('  Model loaded!\n');
}

interface StrapiProduct {
  id: number;
  documentId: string;
  name: string;
  images: { id: number; url: string; name: string }[];
}

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

async function getProductsWithImages(): Promise<StrapiProduct[]> {
  console.log('Fetching products with images...');
  const products: StrapiProduct[] = [];
  let page = 1;

  while (true) {
    const response = await fetch(
      `${STRAPI_URL}/api/products?populate=images&pagination[page]=${page}&pagination[pageSize]=100`,
      { headers: { Authorization: `Bearer ${STRAPI_TOKEN}` } }
    );

    if (!response.ok) {
      throw new Error(`Failed to fetch products: ${response.status}`);
    }

    const data = await response.json();
    if (!data.data || data.data.length === 0) break;

    for (const product of data.data) {
      if (product.images && product.images.length > 0) {
        products.push({
          id: product.id,
          documentId: product.documentId,
          name: product.name,
          images: product.images,
        });
      }
    }

    if (data.data.length < 100) break;
    page++;
  }

  console.log(`  Found ${products.length} products with images\n`);
  return products;
}

async function downloadImage(url: string): Promise<Buffer | null> {
  try {
    const fullUrl = url.startsWith('/') ? `${STRAPI_URL}${url}` : url;
    const response = await fetch(fullUrl);
    if (!response.ok) return null;

    const arrayBuffer = await response.arrayBuffer();
    return Buffer.from(arrayBuffer);
  } catch {
    return null;
  }
}

async function processImage(imageBuffer: Buffer): Promise<Buffer> {
  // First, normalize the image to PNG using Sharp (fixes MIME type issues)
  const normalizedBuffer = await sharp(imageBuffer)
    .png()
    .toBuffer();

  // Convert to Blob for background removal with correct MIME type
  const blob = new Blob([normalizedBuffer], { type: 'image/png' });

  // Remove background
  const resultBlob = await removeBackground(blob);
  const resultBuffer = Buffer.from(await resultBlob.arrayBuffer());

  // Get dimensions of the processed image
  const metadata = await sharp(resultBuffer).metadata();
  const { width = 800, height = 800 } = metadata;

  // Create shadow parameters
  const shadowOffset = 12;
  const shadowBlur = 25;
  const padding = 40;

  // Calculate new canvas size
  const newWidth = width + padding * 2 + shadowOffset;
  const newHeight = height + padding * 2 + shadowOffset;

  // Create a shadow layer (dark semi-transparent rectangle, blurred)
  // We'll composite: white background -> blurred shadow -> original image

  // 1. Create the shadow shape from the alpha channel
  const alphaChannel = await sharp(resultBuffer)
    .extractChannel('alpha')
    .toBuffer();

  // 2. Create shadow: take alpha, make it dark, extend, blur
  const shadow = await sharp(alphaChannel)
    .resize(width, height)
    .extend({
      top: padding,
      bottom: padding + shadowOffset,
      left: padding,
      right: padding + shadowOffset,
      background: { r: 0, g: 0, b: 0, alpha: 0 },
    })
    .blur(shadowBlur)
    .modulate({ brightness: 0 }) // Make it black
    .toBuffer();

  // 3. Create final composite
  const result = await sharp({
    create: {
      width: newWidth,
      height: newHeight,
      channels: 4,
      background: { r: 255, g: 255, b: 255, alpha: 255 },
    },
  })
    .composite([
      // Shadow layer (offset and blurred)
      {
        input: await sharp(shadow)
          .ensureAlpha()
          .modulate({ brightness: 0.3 })
          .toBuffer(),
        top: shadowOffset,
        left: shadowOffset,
        blend: 'multiply',
      },
      // Original image with transparent background
      {
        input: await sharp(resultBuffer)
          .extend({
            top: padding,
            bottom: padding + shadowOffset,
            left: padding,
            right: padding + shadowOffset,
            background: { r: 0, g: 0, b: 0, alpha: 0 },
          })
          .toBuffer(),
        top: 0,
        left: 0,
      },
    ])
    .png()
    .toBuffer();

  return result;
}

async function uploadToStrapi(
  imageBuffer: Buffer,
  filename: string
): Promise<number | null> {
  try {
    const formData = new FormData();
    const blob = new Blob([imageBuffer], { type: 'image/png' });
    formData.append('files', blob, filename);

    const response = await fetch(`${STRAPI_URL}/api/upload`, {
      method: 'POST',
      headers: { Authorization: `Bearer ${STRAPI_TOKEN}` },
      body: formData,
    });

    if (!response.ok) return null;

    const result = await response.json();
    return result?.[0]?.id || null;
  } catch {
    return null;
  }
}

async function updateProductImage(
  documentId: string,
  fileId: number
): Promise<boolean> {
  try {
    const response = await fetch(`${STRAPI_URL}/api/products/${documentId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${STRAPI_TOKEN}`,
      },
      body: JSON.stringify({ data: { images: [fileId] } }),
    });
    return response.ok;
  } catch {
    return false;
  }
}

async function deleteOldFile(fileId: number): Promise<void> {
  try {
    await fetch(`${STRAPI_URL}/api/upload/files/${fileId}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${STRAPI_TOKEN}` },
    });
  } catch {
    // Ignore errors
  }
}

async function main(): Promise<void> {
  console.log('╔════════════════════════════════════════════╗');
  console.log('║   Image Background Removal & Shadow        ║');
  console.log('╚════════════════════════════════════════════╝\n');

  if (!STRAPI_TOKEN) {
    console.error('❌ STRAPI_ADMIN_TOKEN is required');
    process.exit(1);
  }

  // Initialize background removal
  await initBackgroundRemoval();

  // Get products with images
  const products = await getProductsWithImages();

  if (products.length === 0) {
    console.log('No products with images found!');
    return;
  }

  console.log(`Processing ${products.length} product images...\n`);

  let processed = 0;
  let failed = 0;

  for (const product of products) {
    try {
      const image = product.images[0];
      const oldFileId = image.id;

      // Download
      const imageBuffer = await downloadImage(image.url);
      if (!imageBuffer) {
        console.log(`  ✗ ${product.name}: download failed`);
        failed++;
        continue;
      }

      // Process (remove bg + add shadow)
      const resultBuffer = await processImage(imageBuffer);

      // Generate new filename
      const baseName = path.basename(image.name, path.extname(image.name));
      const newFilename = `${baseName}_processed.png`;

      // Upload new image
      const newFileId = await uploadToStrapi(resultBuffer, newFilename);
      if (!newFileId) {
        console.log(`  ✗ ${product.name}: upload failed`);
        failed++;
        continue;
      }

      // Update product
      const success = await updateProductImage(product.documentId, newFileId);
      if (!success) {
        console.log(`  ✗ ${product.name}: link failed`);
        failed++;
        continue;
      }

      // Delete old file
      await deleteOldFile(oldFileId);

      processed++;
      console.log(`  ✓ [${processed}/${products.length}] ${product.name}`);

      // Progress update
      if (processed % 10 === 0) {
        console.log(`\n  Progress: ${processed}/${products.length} (${failed} failed)\n`);
      }

      await delay(100);
    } catch (error) {
      console.log(`  ✗ ${product.name}: ${error}`);
      failed++;
    }
  }

  console.log('\n════════════════════════════════════════════');
  console.log(`✅ Processed: ${processed}`);
  console.log(`❌ Failed: ${failed}`);
  console.log('════════════════════════════════════════════\n');
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
